File: cli/lib/playlist_parser.js
import fs from 'node:fs';
import { stemOf } from './util.js';

// Port of lib/services/playlist_parser.dart
export class PlaylistParser {
  static isInternalPlaylist(name) {
    const n = name.toLowerCase();
    if (n.includes('_removed songs')) return true;
    return ['_unsorted', 'single tracks', '_unsorted_songs'].some((m) => n.includes(m));
  }

  static parseTrackNames(filePath) {
    const songs = [];
    if (!fs.existsSync(filePath)) return songs;
    let content;
    try { content = fs.readFileSync(filePath, 'utf-8'); }
    catch { return songs; }

    const lines = content.split('\n').map((l) => l.trim()).filter((l) => l.length > 0);
    if (lines.length === 0) return songs;
    const isM3u = lines.some((l) => l.includes('#EXTM3U'));

    let i = 0;
    while (i < lines.length) {
      const line = lines[i];
      if (isM3u && line.startsWith('#EXTINF:')) {
        let j = i + 1;
        while (j < lines.length && (lines[j].startsWith('#') || lines[j].length === 0)) j++;
        if (j < lines.length) {
          const pathLine = lines[j];
          const name = pathLine.includes('\\') || pathLine.includes('/')
            ? pathLine.replace(/[\\/]/g, '\\').split('\\').pop()
            : pathLine;
          songs.push(name.replace(/\.[^.]+$/, ''));
          i = j + 1;
        } else {
          i++;
        }
      } else {
        if (!line.startsWith('#')) {
          songs.push(line);
        }
        i++;
      }
    }
    return songs;
  }

  static parseTrackEntries(filePath) {
    const entries = [];
    if (!fs.existsSync(filePath)) return entries;
    let content;
    try { content = fs.readFileSync(filePath, 'utf-8'); }
    catch { return entries; }

    const lines = content.split('\n').map((l) => l.trim()).filter((l) => l.length > 0);
    if (lines.length === 0) return entries;
    const isM3u = lines.some((l) => l.includes('#EXTM3U'));

    let i = 0;
    while (i < lines.length) {
      const line = lines[i];
      if (isM3u && line.startsWith('#EXTINF:')) {
        let j = i + 1;
        while (j < lines.length && (lines[j].startsWith('#') || lines[j].length === 0)) j++;
        if (j < lines.length) {
          entries.push(lines[j]);
          i = j + 1;
        } else {
          i++;
        }
      } else {
        if (!line.startsWith('#')) {
          entries.push(line);
        }
        i++;
      }
    }
    return entries;
  }
}

File: lib/services/playlist_parser.dart
import 'dart:convert';
import 'dart:io';

class PlaylistParser {
  static const _internalMarkers = ['_unsorted', 'single tracks', '_unsorted_songs', '_favorites'];

  static bool isInternalPlaylist(String name) {
    final n = name.toLowerCase();
    if (n.contains('_removed songs')) return true;
    return _internalMarkers.any((m) => n.contains(m));
  }

  static List<String> parseTrackNames(String filePath) {
    final songs = <String>[];
    final file = File(filePath);
    if (!file.existsSync()) return songs;

    String content;
    try {
      content = file.readAsStringSync(encoding: utf8);
    } catch (_) {
      try {
        content = file.readAsStringSync();
      } catch (_) {
        return songs;
      }
    }

    final lines = content.split('\n').map((l) => l.trim()).where((l) => l.isNotEmpty).toList();
    if (lines.isEmpty) return songs;

    final isM3u = lines.any((l) => l.contains('#EXTM3U'));

    int i = 0;
    while (i < lines.length) {
      final line = lines[i];
      if (line.isEmpty) { i++; continue; }

      if (isM3u) {
        if (line.startsWith('#EXTINF:')) {
          int j = i + 1;
          while (j < lines.length && (lines[j].startsWith('#') || lines[j].isEmpty)) {
            j++;
          }
          if (j < lines.length) {
            final pathLine = lines[j];
            final name = pathLine.contains('\\') || pathLine.contains('/')
                ? pathLine.split(RegExp(r'[\\/]')).last
                : pathLine;
            songs.add(_decode(name).replaceAll(RegExp(r'\.\w+$'), ''));
            i = j + 1;
          } else {
            i++;
          }
        } else {
          i++;
        }
      } else {
        if (!line.startsWith('#')) {
          songs.add(line);
        }
        i++;
      }
    }
    return songs;
  }

  static String _decode(String s) {
    // Decode URI-encoded paths (e.g. %E5%A4%A2%E6%83%B3 -> 夢想).
    // Raw paths (incl. literal %) must pass through untouched.
    try {
      return Uri.decodeComponent(s);
    } catch (_) {
      return s;
    }
  }

  static List<String> parseTrackEntries(String filePath) {
    final entries = <String>[];
    final file = File(filePath);
    if (!file.existsSync()) return entries;

    String content;
    try {
      content = file.readAsStringSync(encoding: utf8);
    } catch (_) {
      try {
        content = file.readAsStringSync();
      } catch (_) {
        return entries;
      }
    }

    final lines = content.split('\n').map((l) => l.trim()).where((l) => l.isNotEmpty).toList();
    if (lines.isEmpty) return entries;

    final isM3u = lines.any((l) => l.contains('#EXTM3U'));

    int i = 0;
    while (i < lines.length) {
      final line = lines[i];
      if (line.isEmpty) { i++; continue; }

      if (isM3u) {
        if (line.startsWith('#EXTINF:')) {
          int j = i + 1;
          while (j < lines.length && (lines[j].startsWith('#') || lines[j].isEmpty)) {
            j++;
          }
if (j < lines.length) {
          entries.add(_decode(lines[j]));
          i = j + 1;
          } else {
            i++;
          }
        } else {
          i++;
        }
      } else {
        if (!line.startsWith('#')) {
          entries.add(line);
        }
        i++;
      }
    }
    return entries;
  }
}
